src/inputfetcher.h

```diff
@@ -0,0 +1,237 @@
+            }
+        } while (true);
+    }
+
+    //! Add a batch of outpoints to the queue
+    void Add(std::vector<COutPoint>&& outpoints) noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        if (outpoints.empty()) {
+            return;
+        }
+
+        const auto size{outpoints.size()};
+        {
+            LOCK(m_mutex);
+            m_in_flight_fetches_count += outpoints.size();
+            if (m_outpoints.empty()) {
+                m_outpoints = std::move(outpoints);
+            } else {
+                m_outpoints.insert(m_outpoints.end(), std::make_move_iterator(outpoints.begin()), std::make_move_iterator(outpoints.end()));
+            }
+        }
+
+        if (size == 1) {
+            m_worker_cv.notify_one();
+        } else {
+            m_worker_cv.notify_all();
+        }
+    }
+
+
+public:
+    //! Create a new input fetcher
+    explicit InputFetcher(size_t batch_size, size_t worker_thread_count) noexcept
+        : m_batch_size(batch_size)
+    {
+        m_worker_threads.reserve(worker_thread_count);
+        for (size_t n = 0; n < worker_thread_count; ++n) {
+            m_worker_threads.emplace_back([this, n]() {
+                util::ThreadRename(strprintf("inputfetch.%i", n));
+                Loop();
+            });
+        }
+    }
+
+    // Since this class manages its own resources, which is a thread
+    // pool `m_worker_threads`, copy and move operations are not appropriate.
+    InputFetcher(const InputFetcher&) = delete;
+    InputFetcher& operator=(const InputFetcher&) = delete;
+    InputFetcher(InputFetcher&&) = delete;
+    InputFetcher& operator=(InputFetcher&&) = delete;
+
+
+    //! Fetch all block inputs from db, and insert into cache.
+    void FetchInputs(CCoinsViewCache& cache, const CCoinsViewDB& db, const CBlock& block) noexcept EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        m_db = &db;
+
+        std::vector<COutPoint> buffer{};
+        buffer.reserve(m_batch_size);
+        std::set<Txid> txids{};
+        for (const auto& tx : block.vtx) {
+            if (tx->IsCoinBase()) continue;
+            for (const auto& in : tx->vin) {
+                const auto& outpoint = in.prevout;
+                // If an input references an outpoint from earlier in the
+                // block, it won't be in the cache yet but it also won't be
+                // in the db either.
+                if (txids.contains(outpoint.hash)) {
+                    continue;
+                }
+                if (cache.HaveCoinInCache(outpoint)) {
+                    continue;
+                }
+
+                buffer.emplace_back(outpoint);
+                if (buffer.size() == m_batch_size) {
+                    Add(std::move(buffer));
+                    buffer.clear();
+                    buffer.reserve(m_batch_size);
+                }
+            }
+            txids.insert(tx->GetHash());
+        }
+
+        Add(std::move(buffer));
```

Comment:
do we always have leftovers or will this process the last batch twice (or process an empty one) if the batch happens to be divisible by batch_size?

Reply:
It won't process twice, but it could pass in an empty vector, which is ignored if you look at `Add` implementation.